src/task_monitor/tasks/delete_identities.rs

```diff
@@ -0,0 +1,146 @@
+use std::collections::HashSet;
+use std::sync::Arc;
+
+use anyhow::Result as AnyhowResult;
+use ethers::types::U256;
+use tokio::select;
+use tokio::sync::Notify;
+use tokio::time::{Duration, Instant};
+use tracing::{info, instrument};
+
+use crate::contracts::{IdentityManager, SharedIdentityManager};
+use crate::database::types::DeletionEntry;
+use crate::database::Database;
+use crate::identity_tree::{
+    Canonical, Hash, Intermediate, Latest, TreeVersion, TreeVersionReadOps, TreeWithNextVersion,
+};
+use crate::task_monitor::PendingBatchSubmission;
+use crate::utils::async_queue::{AsyncPopGuard, AsyncQueue};
+
+// It can take up to 40 minutes to bridge the root
+
+pub struct DeleteIdentities {
+    database:                Arc<Database>,
+    identity_manager:        SharedIdentityManager,
+    latest_tree:             TreeVersion<Latest>,
+    deletion_time_interval:  u64,
+    min_deletion_batch_size: usize,
+    wake_up_notify:          Arc<Notify>,
+}
+
+impl DeleteIdentities {
+    pub fn new(
+        database: Arc<Database>,
+        identity_manager: SharedIdentityManager,
+        latest_tree: TreeVersion<Latest>,
+        deletion_time_interval: u64,
+        min_deletion_batch_size: usize,
+        wake_up_notify: Arc<Notify>,
+    ) -> Arc<Self> {
+        Arc::new(Self {
+            database,
+            identity_manager,
+            latest_tree,
+            deletion_time_interval,
+            min_deletion_batch_size,
+            wake_up_notify,
+        })
+    }
+
+    pub async fn run(self: Arc<Self>) -> anyhow::Result<()> {
+        delete_identities(
+            &self.database,
+            &self.identity_manager,
+            &self.latest_tree,
+            self.deletion_time_interval,
+            self.min_deletion_batch_size,
+            self.wake_up_notify.clone(),
+        )
+        .await
+    }
+}
+
+// TODO: we might want to keep track of the last time we completed a deletion
+// batch and make sure that there is a batch run at least once every n days to
+// be compliant
+async fn delete_identities(
+    database: &Database,
+    identity_manager: &IdentityManager,
+    latest_tree: &TreeVersion<Latest>,
+    deletion_time_interval: u64,
+    min_deletion_batch_size: usize,
+    wake_up_notify: Arc<Notify>,
+) -> AnyhowResult<()> {
+    // TODO: do we need to do this for this step?
+    info!("Awaiting for a clean slate");
+    identity_manager.await_clean_slate().await?;
+
+    info!("Starting deletion processor.");
+
+    let deletion_time_interval = Duration::from_secs(deletion_time_interval);
+
+    // TODO: we should track this with persistance to eunsure we are deleting at
```

Comment:
That's a good point. This timeout will probably be very large in prod deployment.

We could estimate the last batch time by fetching the latest root in root history and take it's expiry timestamp?

Reply:
I added a new table that only allows for one row to track the latest deletion.